cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/ConsumerWorkingSetOperations.java

```diff
@@ -0,0 +1,181 @@
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.data2.dataset2.lib.partitioned;
+
+import co.cask.cdap.api.Predicate;
+import co.cask.cdap.api.dataset.lib.PartitionConsumerResult;
+import co.cask.cdap.api.dataset.lib.PartitionDetail;
+import co.cask.cdap.api.dataset.lib.PartitionKey;
+import co.cask.cdap.api.dataset.lib.PartitionedFileSet;
+import co.cask.cdap.api.dataset.lib.partitioned.ConsumerConfiguration;
+import com.google.common.base.Function;
+import com.google.common.collect.Lists;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Defines a set of operations for interacting with a {@link ConsumerWorkingSet}.
+ */
+public class ConsumerWorkingSetOperations {
+  private static final Logger LOG = LoggerFactory.getLogger(ConsumerWorkingSetOperations.class);
+
+  private final ConsumerConfiguration consumerConfiguration;
+  private final PartitionedFileSet partitionedFileSet;
+  private final ConsumerWorkingSet workingSet;
+
+  public ConsumerWorkingSetOperations(ConsumerConfiguration consumerConfiguration,
+                                      PartitionedFileSet partitionedFileSet,
+                                      ConsumerWorkingSet workingSet) {
+    this.consumerConfiguration = consumerConfiguration;
+    this.partitionedFileSet = partitionedFileSet;
+    this.workingSet = workingSet;
+  }
+
+  /**
+   * Consumes partitions from the working set, while marking them as IN_PROGRESS
+   */
+  public List<PartitionDetail> consumePartitions(int consumingLimit) {
+    doExpiry();
+    populateWorkingSet();
+
+    long now = System.currentTimeMillis();
+    List<ConsumablePartition> toConsume = new ArrayList<>();
+
+    List<ConsumablePartition> partitions = workingSet.getPartitions();
+    for (int i = 0; toConsume.size() < consumingLimit && i < partitions.size(); i++) {
+      ConsumablePartition consumablePartition = partitions.get(i);
+      if (!consumablePartition.isInProgress()) {
+        consumablePartition.markInProgress(now);
+        toConsume.add(consumablePartition);
+      }
+    }
+
+    return Lists.transform(toConsume, new Function<ConsumablePartition, PartitionDetail>() {
+      @Override
+      public PartitionDetail apply(ConsumablePartition input) {
+        return partitionedFileSet.getPartition(input.getPartitionKey());
+      }
+    });
+  }
+
+  /**
+   * Fetches additional partitions from the PartitionedFileSet and adds them to the ConsumerWorkingSet.
+   */
+  private void populateWorkingSet() {
+    int numToPopulate = consumerConfiguration.getMaxWorkingSetSize() - workingSet.getPartitions().size();
```

Comment:
what happens if the working set had reached its max size? Then this number is 0. Doe pfs.consumerPartitions handle that correctly?